content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+  return ErrCouldNotOpen
+}
+
+// package bar
+
+if err := foo.Open(); err != nil {
+  if errors.Is(err, foo.ErrCouldNotOpen) {
+    // handle
+  } else {
+    panic("unknown error")
+  }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+If you have an error that clients may need to detect, and you would like to add more information to it (e.g., it is not a static string), then you should use a custom type.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func open(file string) error {
+  return fmt.Errorf("file %q not found", file)
+}
+
+func use() {
+  if err := open("testfile.txt"); err != nil {
+    if strings.Contains(err.Error(), "not found") {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td><td>
+
+```go
+type errNotFound struct {
+  file string
+}
+
+func (e errNotFound) Error() string {
+  return fmt.Sprintf("file %q not found", e.file)
+}
+
+func open(file string) error {
+  return errNotFound{file: file}
+}
+
+func use() {
+  if err := open("testfile.txt"); err != nil {
+    if _, ok := err.(errNotFound); ok {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Be careful with exporting custom error types directly since they become part of the public API of the package. It is preferable to expose matcher functions to check the error instead.
+
+```go
+// package foo
+
+type errNotFound struct {
+  file string
+}
+
+func (e errNotFound) Error() string {
+  return fmt.Sprintf("file %q not found", e.file)
+}
+
+func IsNotFoundError(err error) bool {
+  _, ok := err.(errNotFound)
+  return ok
+}
+
+func Open(file string) error {
+  return errNotFound{file: file}
+}
+
+// package bar
+
+if err := foo.Open("foo"); err != nil {
+  if foo.IsNotFoundError(err) {
+    // handle
+  } else {
+    panic("unknown error")
+  }
+}
+```
+
+### Error Wrapping
+
+There are three main options for propagating errors if a call fails:
+
+* Return the original error if there is no additional context to add and you want to maintain the original error type.
+* Add context using [`"pkg/errors".Wrap`] so that the error message provides more context and [`"pkg/errors".Cause`] can be used to extract the original error.
+* Use [`fmt.Errorf`] if the callers do not need to detect or handle that specific error case.
+
+It is recommended to add context where possible so that instead of a vague error such as "connection refused", you get more useful errors such as "call service foo: connection refused".
+
+See also [Don't just check errors, handle them gracefully].
+
+[`"pkg/errors".Cause`]: https://godoc.org/github.com/pkg/errors#Cause
+[Don't just check errors, handle them gracefully]: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully
+
+### Handle Type Assertion Failures
+
+The single return value form of a [type assertion] will panic on an incorrect type. Therefore, always use the "comma ok" idiom.
```

Comment:
Or a type switch, depending on the scenario. But perhaps it dilutes the point.

Reply:
I think so yeah 😄 I think it's fine like this.